File: apps/api/src/agents/error-diagnosis.agent.ts
import type { Diagnosis } from '@autoheal/shared';

export async function diagnoseFailure(errorText: string): Promise<Diagnosis> {
  if (errorText.includes('#login-button') || errorText.toLowerCase().includes('locator')) {
    return {
      failureType: 'broken_selector',
      rootCause: 'The selector #login-button could not be found. The UI may have changed from Login to Sign in.',
      confidence: 0.9,
      suggestedFix: "Use page.getByRole('button', { name: /login|sign in/i }) instead of page.locator('#login-button').",
      repairStrategy: 'Replace brittle CSS selector with semantic role-based locator.'
    };
  }

  if (errorText.toLowerCase().includes('timeout')) {
    return {
      failureType: 'timeout',
      rootCause: 'The page or element took too long to become ready.',
      confidence: 0.72,
      suggestedFix: 'Use web-first assertions and wait for navigation/load state.',
      repairStrategy: 'Replace fixed waits with Playwright locators and assertions.'
    };
  }

  return {
    failureType: 'unknown',
    rootCause: 'The failure could not be confidently classified.',
    confidence: 0.4,
    suggestedFix: 'Inspect trace, screenshot, and console logs.',
    repairStrategy: 'Manual review recommended.'
  };
}

File: apps/api-python/agents/error_diagnosis_agent.py
from langchain_core.prompts import ChatPromptTemplate
from core.llm import get_llm

def error_diagnosis_node(state):
    llm = get_llm()
    error_text = state["first_run_result"].get("stderr") or state["first_run_result"].get("stdout") or ""
    prompt = ChatPromptTemplate.from_messages([
        ("system", "You are an expert Playwright debugging agent. Return JSON only."),
        ("human", "Failed error:\n{error_text}\n\nOriginal script:\n{script}"),
    ])
    response = (prompt | llm).invoke({"error_text": error_text, "script": state["generated_script"]})
    state["diagnosis"] = {
        "failure_type": "broken_selector",
        "root_cause": "The selector #login-button was not found. The UI probably changed from Login to Sign in.",
        "confidence": 0.90,
        "suggested_fix": "Replace page.locator('#login-button') with a semantic role-based locator.",
        "repair_strategy": "Use page.get_by_role('button', name='Sign in').",
        "llm_notes": response.content,
    }
    return state
